contracts/test/StakeTableInvariant.t.sol

```diff
@@ -0,0 +1,341 @@
+
+        LightClient.LightClientState memory genesis = LightClient.LightClientState({
+            viewNum: 0,
+            blockHeight: 0,
+            blockCommRoot: BN254.ScalarField.wrap(1),
+            feeLedgerComm: BN254.ScalarField.wrap(1),
+            stakeTableBlsKeyComm: BN254.ScalarField.wrap(1),
+            stakeTableSchnorrKeyComm: BN254.ScalarField.wrap(1),
+            stakeTableAmountComm: BN254.ScalarField.wrap(1),
+            threshold: 10
+        });
+        uint32 numBlocksPerEpoch = 10;
+        uint64 churnRate = 10;
+        lightClientContract = new LightClientTest(genesis, numBlocksPerEpoch);
+        stakeTable = new S(address(token), address(lightClientContract), churnRate);
+        handler =
+            new StakeTableHandlerTest(stakeTable, exampleTokenCreator, token, lightClientContract);
+
+        // Only test the handler
+        targetContract(address(handler));
+    }
+
+    function invariant_BalancesAreConsistent() external {
+        uint256 totalBalanceUsers = 0;
+        for (uint256 i = 0; i < users.length; i++) {
+            totalBalanceUsers += token.balanceOf(users[i]);
+        }
+        uint256 balanceStakeTable = token.balanceOf(address(stakeTable));
+        uint256 tokenCreatorBalance = token.balanceOf(address(exampleTokenCreator));
+        assertEq(totalBalanceUsers + balanceStakeTable + tokenCreatorBalance, INITIAL_BALANCE);
+    }
+
+    function invariant_Register() external {
+        // Here we check that the queue state is updated in a consistent manner with the output
+        // of nextExitEpoch.
+        if (handler.registrationCalledAtLeastOnce()) {
+            assertEq(
+                handler.stakeTableFirstAvailableRegistrationEpoch(),
+                handler.nextRegistrationEpochBefore()
+            );
+            assertEq(
+                handler.stakeTableNumPendingRegistrations(),
+                handler.pendingRegistrationsBefore() + 1
+            );
+        } else {
+            assertEq(handler.stakeTableFirstAvailableRegistrationEpoch(), 0);
+            assertEq(handler.stakeTableNumPendingRegistrations(), 0);
+        }
+    }
+
+    function invariant_RequestExit() external {
+        // Here we check that the queue state is updated in a consistent manner with the output
+        // of nextExitEpoch.
+        if (handler.requestExitCalledAtLeastOnce()) {
+            assertGe(handler.stakeTableFirstAvailableExitEpoch(), handler.currentEpoch() + 1);
+            assertGe(handler.stakeTableNumPendingExits(), 1);
+        } else {
+            assertEq(handler.stakeTableFirstAvailableExitEpoch(), 0);
+            assertEq(handler.stakeTableNumPendingExits(), 0);
+        }
+    }
+
+    function invariant_Queue() external {
+        // Global invariants
+        assertLe(stakeTable.numPendingRegistrations(), stakeTable.maxChurnRate());
+        assertLe(stakeTable.numPendingExits(), stakeTable.maxChurnRate());
+    }
```

Comment:
what about: keep a counter for the total registration & total exit in the handler, and make sure our `total reg = churn rate * nextEpoch + numPending`?

Reply:
This is true only if we don't skip epochs.

Reply:
can we also document the number of skipped epoch?

Reply:
I don't think that would work either because an epoch can be skipped when the first available epoch is partially filled. So we cannot derive a clean equation between the total number of registrations and the number of epochs (skipped and not skipped) and the churn rate.